src/bson.c

```diff
@@ -964,71 +1000,63 @@ PHP_FUNCTION(fromPHP)
 }
 /* }}} */
 
+static void apply_classname_to_state(const char *classname, int classname_len, php_phongo_bson_typemap_types *type, zend_class_entry **type_ce TSRMLS_DC)
+{
+	if (!strcasecmp(classname, "array")) {
+		*type = PHONGO_TYPEMAP_NATIVE_ARRAY;
+		*type_ce = NULL;
+	} else if (!strcasecmp(classname, "stdclass") || !strcasecmp(classname, "object")) {
+		*type = PHONGO_TYPEMAP_NATIVE_OBJECT;
+		*type_ce = NULL;
+	} else {
+		zend_class_entry *found_ce = zend_fetch_class(classname, classname_len, ZEND_FETCH_CLASS_AUTO|ZEND_FETCH_CLASS_SILENT TSRMLS_CC);
+
+		if (!found_ce) {
+			phongo_throw_exception(PHONGO_ERROR_INVALID_ARGUMENT TSRMLS_CC, "Class %s does not exist", classname);
+		} else if (!IS_CLASS_INSTANTIATABLE(found_ce)) {
+			phongo_throw_exception(PHONGO_ERROR_INVALID_ARGUMENT TSRMLS_CC, "Class %s is not instantiatable", classname);
+		} else if (!instanceof_function(found_ce, php_phongo_unserializable_ce TSRMLS_CC)) {
+			phongo_throw_exception(PHONGO_ERROR_INVALID_ARGUMENT TSRMLS_CC, "Class %s does not implement %s\\Unserializable", classname, BSON_NAMESPACE);
+		} else {
+			*type = PHONGO_TYPEMAP_CLASS;
+			*type_ce = found_ce;
+		}
+	}
+}
+
 void php_phongo_bson_typemap_to_state(zval *typemap, php_phongo_bson_typemap *map TSRMLS_DC)
 {
 	if (typemap) {
-		char                  *classname;
-		int                    classname_len;
-		zend_bool              classname_free = 0;
+		char      *classname;
+		int        classname_len;
+		zend_bool  classname_free = 0;
 
 		classname = php_array_fetchl_string(typemap, "array", sizeof("array")-1, &classname_len, &classname_free);
 		if (classname_len) {
-			if (!strcasecmp(classname, "array")) {
-				map->array_type = PHONGO_TYPEMAP_NATIVE_ARRAY;
-			} else if (!strcasecmp(classname, "stdclass") || !strcasecmp(classname, "object")) {
-				map->array_type = PHONGO_TYPEMAP_NATIVE_OBJECT;
-			} else {
-				zend_class_entry *array_ce = zend_fetch_class(classname, classname_len, ZEND_FETCH_CLASS_AUTO TSRMLS_CC);
-				map->array_type = PHONGO_TYPEMAP_CLASS;
-
-				if (instanceof_function(array_ce, php_phongo_unserializable_ce TSRMLS_CC)) {
-					map->array = array_ce;
-				}
-			}
-			if (classname_free) {
-				efree(classname);
-			}
+			apply_classname_to_state(classname, classname_len, &map->array_type, &map->array TSRMLS_CC);
+		}
+		if (classname_free) {
```

Comment:
Based on `char *php_array_zval_to_string(zval *z, int *plen, zend_bool *pfree)` in `php_array_api.h`, `pfree` is only set to zero if `zval *z` is `NULL` or a string, in which case its direct value is returned. I believe I encountered a memory leak in the test where I passed in `null` to each type map key. This uses `STR_EMPTY_ALLOC()`, which might return an interned or allocated, empty string (which is why I started using `str_efree()` on it, since `efree()` isn't aware of interned strings).